spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/GcpCloudSqlAutoConfiguration.java

```diff
@@ -19,144 +19,233 @@
+					new DefaultCloudSqlJdbcInfoProvider(gcpCloudSqlProperties, DatabaseType.MYSQL);
+
+			if (LOGGER.isInfoEnabled()) {
+				LOGGER.info("Default " + DatabaseType.MYSQL.name()
+						+ " JdbcUrl provider. Connecting to "
+						+ defaultProvider.getJdbcUrl() + " with driver "
+						+ defaultProvider.getJdbcDriverClass());
+			}
+
+			return defaultProvider;
 		}
 
-		setCredentialsProperty();
+	}
+
+	/**
+	 * The PostgreSql Configuration for the {@link DefaultCloudSqlJdbcInfoProvider}
+	 * based on the {@link DatabaseType#POSTGRESQL}.
+	 */
+	@ConditionalOnClass({ com.google.cloud.sql.postgres.SocketFactory.class, org.postgresql.Driver.class })
+	@ConditionalOnMissingBean(CloudSqlJdbcInfoProvider.class)
+	static class PostgreSqlJdbcInfoProviderConfiguration {
+
+		@Bean
+		public CloudSqlJdbcInfoProvider defaultPostgreSqlJdbcInfoProvider(GcpCloudSqlProperties gcpCloudSqlProperties) {
+			CloudSqlJdbcInfoProvider defaultProvider =
+					new DefaultCloudSqlJdbcInfoProvider(gcpCloudSqlProperties, DatabaseType.POSTGRESQL);
+
+			if (LOGGER.isInfoEnabled()) {
+				LOGGER.info("Default " + DatabaseType.POSTGRESQL.name()
+						+ " JdbcUrl provider. Connecting to "
+						+ defaultProvider.getJdbcUrl() + " with driver "
+						+ defaultProvider.getJdbcDriverClass());
+			}
+
+			return defaultProvider;
+		}
 
-		return properties;
 	}
 
 	/**
-	 * Set credentials to be used by the Google Cloud SQL socket factory.
-	 *
-	 * <p>The only way to pass a {@link CredentialFactory} to the socket factory is by passing a
-	 * class name through a system property. The socket factory creates an instance of
-	 * {@link CredentialFactory} using reflection without any arguments. Because of that, the
-	 * credential location needs to be stored somewhere where the class can read it without
-	 * any context. It could be possible to pass in a Spring context to
-	 * {@link SqlCredentialFactory}, but this is a tricky solution that needs some thinking
-	 * about.
-	 *
-	 * <p>If user didn't specify credentials, the socket factory already does the right thing by
-	 * using the application default credentials by default. So we don't need to do anything.
+	 * The Configuration to populated {@link DataSourceProperties} bean
+	 * based on the cloud-specific properties.
 	 */
-	private void setCredentialsProperty() {
-		File credentialsLocationFile;
-
-		try {
-			// First tries the SQL configuration credential.
-			if (this.gcpCloudSqlProperties != null
-					&& this.gcpCloudSqlProperties.getCredentials() != null) {
-				credentialsLocationFile =
-						this.gcpCloudSqlProperties.getCredentials().getLocation().getFile();
+	@Configuration
+	@Import({ GcpCloudSqlAutoConfiguration.AppEngineJdbcInfoProviderConfiguration.class,
+			GcpCloudSqlAutoConfiguration.MySqlJdbcInfoProviderConfiguration.class,
+			GcpCloudSqlAutoConfiguration.PostgreSqlJdbcInfoProviderConfiguration.class })
+	static class CloudSqlDataSourcePropertiesConfiguration {
```

Comment:
Not sure if I like this more than the three classes before...

Reply:
Please, read the commit comments.
Otherwise we have to copy/paste all the conditions on the `GcpCloudSqlAutoConfiguration` to all those external classes.

We can't trigger, for example, `MySqlJdbcInfoProviderAutoConfiguration` if we don't have `DataSource.class` on CP. We can't do that when we don't meet `spring.cloud.gcp.sql.enabled` condition and also we have to trigger `@EnableConfigurationProperties(GcpCloudSqlProperties.class)` because this `MySqlJdbcInfoProviderAutoConfiguration` has to be performed before `cloudSqlDataSourceProperties`.

That's not about "like - don't like". That's more about the reasonable ceiling of copy/paste and how often we are going to miss add extra condition to those external classes.

OTOH, please, take a look into Spring Boot auto-configuration classes to understand how design can be and what would be better to avoid.

Does it make sense?

Reply:
After looking at this with fresh eyes, it makes sense to me.